File: Exercise_2.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/jump-game-ii/

class Solution {
public:
    int jump(vector<int>& nums) {
        int jmp=0, mj=0, count=0, n=nums.size();
        for(int i=0;i<n-1;i++) {
            jmp = max(jmp,i+nums[i]);
            if(jmp >= n-1) {
                return count+1;   
            }
            
            if(i==mj) {
                mj = jmp;
                count++;
                jmp = 0;
            }
        }
        
        return count;
    }
};

File: Exercise_2.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int c = 0, ne = 0;
        int j =0;
        for(int i =0; i<n-1; ++i){
            ne = Math.max(ne, i+nums[i]);
            if(i == c){
                ++j;
                c =ne;
            }
       }

       return j;

    }
}
//tc = O(n)
//sc = O(1)
